Ecs-1

benny-n opened the pull request:
Details:
- Structs and traits:
  - Component - abstract trait (interface). Any object that wants to be a component needs to ```impl``` this trait.
  - Entity - an abstract object with a unique id. Components may be attached to an entity via scene API.
  - Scene - manager of entities and component pools (this is subject to change, I will open an issue regarding this matter soon). Provides API for attaching components to entities, and holds mappings between component types and their respective component pool. 
  Each member in a component pool consists of a specific component, and an entity ID to indicate which entity holds this component.
- Scene API:
  - Create entity.
  - Attach component to an entity.
  - Get attached component from an entity (as mutable and immutable).
  - Removing components and entities is **yet to be implemented**.
- Macros:
  - Created a #[derive Component] macro - if a struct wants to be a Component, we can simply add ```#[derive Component]``` above the struct's definition and it will automatically implement the component trait thanks to the macro.
  - Created a new crate for the macro definition, and changed the Cargo.toml file to hold a workspace made of 2 crates: gear and component_macro. 
  The reason for this was:  "When creating procedural macros, the definitions must reside in their own crate with a special crate type. This is for complex technical reasons that we hope to eliminate in the future." - _https://doc.rust-lang.org/book/ch19-06-macros.html_
- Tests:
  - Rearranged the tests - instead of having one single tests.rs file, created a folder of tests in each module and implemented module-specific tests in said folder.

benny-n committed:
rearrange tests + begin implementing ecs

benny-n committed:
refactor get_component_type + add component pool size test + remove .toml dependencies

benny-n committed:
add #[derive(Component)] macro

benny-n committed:
add "todo" for remove component

orenkoren commented on component_macro/src/lib.rs:
```diff
@@ -0,0 +1,30 @@
+pub fn component_macro_derive(input: TokenStream) -> TokenStream {
+    // Construct a representation of Rust code as a syntax tree
+    // that we can manipulate
+    let ast = syn::parse(input).unwrap();
```
meaningful variable names pls

orenkoren replied:
i won't duplicate this comment but please go over the entire PR and try to improve the variable names

benny-n replied:
this code is actually taken from the official rust book in the section about derive macros:
![image](https://user-images.githubusercontent.com/66024037/126869932-ffd05b8f-b78c-4565-9655-db6340dfcd8a.png)
ast stands for abstract syntax tree, a common abbreviation in rust macros.
syn is a module, and therefore I have no control over its name.
other than that, I think the variable names are pretty meaningful.

orenkoren commented on src/ecs/scene.rs:
```diff
@@ -0,0 +1,94 @@
+    component_name_to_type_id : HashMap<String, ComponentTypeId>,
+    component_pools : HashMap<ComponentTypeId, ComponentPool>,
+}
+impl Scene{
```
can we make this a singleton? there's only one scene in this context

benny-n replied:
should scene really be a singleton? how would we support multiple scenes?
maybe the managment of components and entities shouldn't be the scene's responsibility, but then if we make a singleton component and entity managers, they would have to hold yet another mapping from scenes to component pools and entities.
to me it seems like too much, but we should discuss this in detail and maybe change the design.

orenkoren replied:
you're using scene as if it's an entity manager. and there's only one entity manager.

benny-n replied:
how would you handle the case of 2 or more scenes?

IliaFeldgun replied:
Unless you guys think two scenes can't exists side by side, I am with @benny-n .

orenkoren commented on src/ecs/scene.rs:
```diff
@@ -0,0 +1,94 @@
+        let component_id = &self.get_component_type_id::<T>();
+
+        let component_pool = self.component_pools.get_mut(component_id);
+        match component_pool{
```
i think it would be more readable if you extracted every match block into it's own private func

benny-n replied:
for now the ```Err```s we have only print messages, but maybe in the future we would want to pass error objects to the caller methods. that being said, even if I extract the match blocks into functions, the functions would have to return ```Result<T,E>``` and to handle this I would have to use match again. 
in short, I don't think we can avoid matching here. 
let me know if you have an idea for an implementation for what you suggested.

orenkoren replied:
can't you just pull the whole match block to a function?

benny-n replied:
I can, but then I would still need to match on that new function's return value (explained why in the 1st comment), so the code will look exactly the same.

IliaFeldgun commented on src/ecs/scene.rs:
```diff
@@ -0,0 +1,94 @@
+pub struct Scene{
+    component_type_counter : i64,
+    entity_id_counter : i64,
+    entities : Vec<Entity>,
```
Why not 
```suggestion
    entities : HashMap<EntityId, Entity>,
```

benny-n replied:
indeed this should probably change to hashmap

benny-n committed:
refactor according to PR review

- refactor "scene" to singleton entity manager
- change entities vec to hashmap
- run tests with only 1 thread, so avoid poisoning the entity manager singleton

benny-n committed:
small refactor of "assign" fn
